File: apps/mobile/android/app/src/main/java/com/beacon/bluetooth/mapping/ByteArrayMapper.kt
package com.beacon.bluetooth.mapping

/**
 * Converts between the bridge's byte representation (`number[]` of unsigned values) and the
 * `ByteArray` the Android GATT APIs take. Pure Kotlin so the validation is unit tested on the
 * JVM; the module unpacks the `ReadableArray` into the list this consumes.
 */
object ByteArrayMapper {

    /**
     * The payload as bytes, or null when any value is not an integer in 0..255 (PROJECT.md 16:
     * "reject bytes above 255"). JavaScript numbers arrive as doubles, so fractional values and
     * NaN are rejected as well.
     */
    fun fromDoubles(values: List<Double>): ByteArray? {
        val bytes = ByteArray(values.size)
        for ((index, value) in values.withIndex()) {
            if (value.isNaN() || value < 0.0 || value > 255.0 || value != Math.floor(value)) return null
            bytes[index] = value.toInt().toByte()
        }
        return bytes
    }

    /** The bytes as unsigned integers 0..255, the form `readCharacteristic` resolves with. */
    fun toUnsignedInts(bytes: ByteArray): List<Int> = bytes.map { it.toInt() and 0xFF }
}

File: apps/mobile/ios/BeaconBluetooth/Mapping/ByteArrayMapper.swift
import Foundation

/// Converts characteristic values between the bridge's number arrays and `Data`. Pure so the
/// range validation PROJECT.md 16 asks for ("reject bytes above 255") is unit tested.
public enum ByteArrayMapper {

  /// The unsigned bytes JavaScript sent, or nil when any element is not a whole number in
  /// `0...255`. React Native delivers every JavaScript number as a double-backed `NSNumber`,
  /// so a fractional, negative, oversized or non-finite value is caught here rather than
  /// truncated on its way into `Data`.
  public static func bytes(from numbers: [NSNumber]) -> [UInt8]? {
    var bytes: [UInt8] = []
    bytes.reserveCapacity(numbers.count)
    for number in numbers {
      let value = number.doubleValue
      guard value >= 0, value <= 255, value.rounded() == value else { return nil }
      bytes.append(UInt8(value))
    }
    return bytes
  }

  /// A characteristic value as the unsigned byte array `readCharacteristic` resolves with.
  public static func numbers(from data: Data) -> [NSNumber] {
    data.map { NSNumber(value: $0) }
  }
}
